# Importando as bibliotecas

In [0]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


# Importando a base de dados


In [0]:
dados = pd.read_csv('/content/Mission_Prediction_Dataset.csv')

In [0]:
dados.head()

age  sex  cp  trestbps  chol  fbs  ...  exang  oldpeak  slope  ca  thal  target
0   63    1   3       145   233    1  ...      0      2.3      0   0     1       1
1   37    1   2       130   250    0  ...      0      3.5      0   0     2       1
2   41    0   1       130   204    0  ...      0      1.4      2   0     2       1
3   56    1   1       120   236    0  ...      0      0.8      2   0     2       1
4   57    0   0       120   354    0  ...      1      0.6      2   0     2       1

[5 rows x 14 columns]

Como estou usando o mesmo dataset, eu irei me abster de repetir toda a avaliação qualitativa sobre os dados. Nesse modelo para não ficar redundante.

# Divisão Base de treinamento X Base de teste

In [0]:
X = dados.iloc[:, :-1].values
Y = dados.iloc[:, -1:].values

In [0]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.3)

# Construção e treinamento do modelo

In [0]:
from sklearn.tree import DecisionTreeClassifier

In [0]:
dtree = DecisionTreeClassifier()

In [0]:
dtree.fit(X_train, Y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [0]:
pred = dtree.predict(X_test)

In [0]:
from sklearn.metrics import classification_report, confusion_matrix

In [0]:
print(classification_report(Y_test, pred))

              precision    recall  f1-score   support

           0       0.78      0.74      0.76        43
           1       0.78      0.81      0.80        48

    accuracy                           0.78        91
   macro avg       0.78      0.78      0.78        91
weighted avg       0.78      0.78      0.78        91



In [0]:
print(confusion_matrix(Y_test, pred))

[[32 11]
 [ 9 39]]


Percebe-se que a accuracy é 78% para uma única árvore de decisão.

Eu tentarei fazer de uma forma estocástica a seguir.

In [0]:
from sklearn.ensemble import RandomForestClassifier

In [0]:
rfc = RandomForestClassifier(n_estimators=500)

In [0]:
rfc.fit(X_train, Y_train)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  """Entry point for launching an IPython kernel.


RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=500,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [0]:
rfc_pred = rfc.predict(X_test)

In [0]:
print(classification_report(Y_test,rfc_pred))

              precision    recall  f1-score   support

           0       0.89      0.79      0.84        43
           1       0.83      0.92      0.87        48

    accuracy                           0.86        91
   macro avg       0.86      0.85      0.86        91
weighted avg       0.86      0.86      0.86        91



In [0]:
print(confusion_matrix(Y_test,rfc_pred))

[[34  9]
 [ 6 42]]


Vê-se que um floresta de decisão possui uma accuracy superior a de uma única árvore sozinha, chegando a 86%.